### PR TITLE
RFC0434: add jwe_suite

### PR DESCRIPTION
**Pull request by llorllale**
Signed-off-by: George Aristy <george.aristy@securekey.com>

### CODE REVIEW

**Review by Baha-sk — APPROVED** [review]

**Review by kdenhartog — COMMENTED** [review]
We're not currently using the OOB protocol for initial setup and only are using DIDCommV2. How would we be expected to convey this?

**Comment by llorllale** [comment]
@kdenhartog 

> We're not currently using the OOB protocol for initial setup and only are using DIDCommV2. How would we be expected to convey this?

I understand we are punting the switch from DIDComm V1 to V2.

cc @swcurran to see if Stephen has any thoughts on this.

**Comment by swcurran** [comment]
@kdenhartog -- how are you doing the initial setup?

**Comment by kdenhartog** [comment]
> @kdenhartog -- how are you doing the initial setup?

Given right now we're only doing issuance/verification of VCs and DID Auth through OIDC provider we are able to use the 0-RTT functionality to build a DIDComm connection while issuing a VC / Verifying a VC during an OIDC interaction

**Comment by llorllale** [comment]
@kdenhartog 

> > @kdenhartog -- how are you doing the initial setup?
> 
> Given right now we're only doing issuance/verification of VCs and DID Auth through OIDC provider we are able to use the 0-RTT functionality to build a DIDComm connection while issuing a VC / Verifying a VC during an OIDC interaction

Let me re-phrase the question: how are the two parties negotiating (JWE?) encryption parameters?

Should we assume you're using [published OIDC provider metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata)?

**Comment by kdenhartog** [comment]
> @kdenhartog
> 
> > > @kdenhartog -- how are you doing the initial setup?
> > 
> > 
> > Given right now we're only doing issuance/verification of VCs and DID Auth through OIDC provider we are able to use the 0-RTT functionality to build a DIDComm connection while issuing a VC / Verifying a VC during an OIDC interaction
> 
> Let me re-phrase the question: how are the two parties negotiating (JWE?) encryption parameters?
> 
> Should we assume you're using [published OIDC provider metadata](https://openid.net/specs/openid-connect-discovery-1_0.html#ProviderMetadata)?

We support all types defined in DIDCommV2 whether optional or required so don't need to negotiate. This is a luxury afforded to us because we're only inter-oping with ourselves at the moment.

In talking with my team we see 3 potential options that we're weighing at the moment.

1. Include the discovery information in the DID Document.
  - this works well for did peer method, but presents challenges for basically every other method type so it probably won't be a good option.

2. Include the discovery information in a public metadata point similar to OIDC
  - This offers a layer of indirection that adds some form of privacy, but because the metadata point still would need to be published in some way (requires effectively the same amount of disclosure as the DID Document - e.g. public ledger DID means public metadata endpoint) we're not certain that the privacy would hold. Furthermore, it offers another point that has to be supported for highly available agents / architected around for semi available agents in order to make discovery work. Additionally, we're not confident in the willingness of others to adopt this method as well at this point so likely won't go this path.
  
3. Support an optional discovery JWM message similar to OOB
 - This is the path that makes the most sense to me at this point because it maps well to the same processes of how you'd discover the did document data with a 0-RTT system while also allowing the ability for an explicit discovery of supported protocols and other data related such that it could be passed at any time rather than only at the beginning. This is important if you want to update your support for data without having to restart the connection. Importantly though, it's not the required first message, but can act as a fall back message if a 0-RTT message is not understood. Particularly what I've been thinking about with this is that the majority of enterprise cases don't allow for any inherent negotiation and instead it's more of a "take it or leave it approach" because of the need to operate at scale. However, when you're interacting in a peer-to-peer environment the assumption that a connection will be set and then doing some form of logic beyond a connection later on seems far more logical and so if you wanted to you could send the discovery message as the first message, but it shouldn't be required in our view. This is the thing that makes a fair amount of sense to us at the moment, but we're open to considering options 1 or 2 / other options if there's strong reasons for doing so.
 
 In summary, right now we don't need to do anything like this because all of our flows operate in a 0-RTT flow and we're supporting all forms of DIDCommV2 encryption. We also recognize the need for this in a subset of the use case we're considering in the future, but have deferred on this to this point because we see it as something that we can upgrade to overtime. Additionally, because we've opted to stick with DIDCommV2 only, we've not spent much time considering the usage of the OOB protocol to date, but want. to make sure that what we consider will align with everyone when Aries begins to converge on DIDCommV2.

**Comment by llorllale** [comment]
@kdenhartog 

- Option 1 does not sound appealing to me because it would complicate even further the encoding of ledger-less DIDs. We are already struggling to come up with ways to "just" encode more than one key in a way that makes sense.
- Option 2 puts an onerous burden on agents that are not always online - just like you said

As for Option 3 - agree it should not be required. In fact, this proposal pretty much enables this use case while still not invalidating the use of the Features Discovery protocol after connection establishment.

Perhaps we can delay discussion about the exact details of how the JWE parameters will be conveyed and instead come to agree on **where** in the protocols this should happen? My proposal is:

* allow the sender of an OOB message to declare their supported JWE parameters _somehow_ (precedent: already disclosing supported handshake protocols)
* also allow an agent to discover the JWE parameters via the discovery protocol

I think the first point is more pressing - we're already discussing adoption of RFC0334 which would smooth the transition to DIDComm V2.

@swcurran @TelegramSam @dhh1128 thoughts?
